mypy/checker.py

```diff
@@ -1901,8 +1901,20 @@ def check_assignment(self, lvalue: Lvalue, rvalue: Expression, infer_lvalue_type
                 rvalue_type = self.expr_checker.accept(rvalue)
                 if not inferred.is_final:
                     rvalue_type = remove_instance_last_known_values(rvalue_type)
+                self.check_forbidden_inference_types(rvalue_type, rvalue)
                 self.infer_variable_type(inferred, lvalue, rvalue_type, rvalue)
 
+    def check_forbidden_inference_types(self, rvalue_type: Type, rvalue: Context) -> None:
+        if isinstance(rvalue_type, Instance):
+            if rvalue_type.type.fullname() == 'builtins.list':
+                list_arg_type = rvalue_type.args[0]
+                if isinstance(list_arg_type, Instance) and list_arg_type.type.fullname() == 'builtins.object':
+                    self.msg.forbidden_inference_of_object(rvalue_type, rvalue)
+            elif rvalue_type.type.fullname() == 'builtins.dict':
+                dict_value_type = rvalue_type.args[1]
+                if isinstance(dict_value_type, Instance) and dict_value_type.type.fullname() == 'builtins.object':
+                    self.msg.forbidden_inference_of_object(rvalue_type, rvalue)
```

Comment:
I think we should not emit these errors if `--allow-untyped-globals` is set to true and we are not at function scope (especially if we change the message to be just `Need type annotation ...`).